etc/docs/Design.md

```diff
@@ -0,0 +1,125 @@
+# Things to cover
+
+## API
+
+### High level Library API
+
+The library API should provide the following methods to the clients:
+
+* Initialize an executor instance with the following settings
+  * Limits for resources
+    
+
+* Execute a command with the supplied arguments, that returns a pointer to a `Command` object that can be later used to manipulate the underlying process
+
+### GRPC API definition
+
+The GRPC API defines the following methods:
+
+* ExecuteCommand: Starts a command with limited resources in its own namespace with the supplied arguments
+* StopExecution : Stop a certain execution, issues a SIGTERM than after a configured amount of seconds issues a SIGKILL
+* ListExecutions: List all the executions the connected client has rights to
+* GetExecution: Gets metadata about a certain execution defined by an execution ID
+* GetOutStream: Get a stream of standard output (stdout & stderr) of a certain execution
+
+To read about the exact details of the API you can look at [the proto file](api.proto). The server will return UUIDs as ExecutionId that will be then used to later identify a certain execution. The reason for not using PID as an ExecutionId is because PIDs could be reused which could lead to unexpected behaviour from the application
+
+### Streaming data
+
+In order to be able to stream and also send the full output of the processes we need to store the output of this in memory for simplicty. See [Tradeoffs](#tradeoffs) section for information about why this solution should not be used in production  
+
+## Security
+
+### Cipher suite
+
+In order to have the latest and strongest transport encryption TLS 1.3 should be utilized. The cipher suite TLS_AES_128_GCM_SHA256 is the mandatory cipher suite for TLS 1.3, and it provides sufficiently strong encryption there for it will be used by the server
+The decision to ultimately use TLS 1.3 is due to the fact that the client and the server controlled by the same entity, therefore it's easy to enforce the usage of the latest version
+
+### Authentication
+
+### mTLS
+
+For the mTLS the following certificates will be generated by a generate-certs shell script to enable easy testing of the application (see the Tradeoff section for more details):
+
+* A root CA certificate to sign client and server CSR-s 
+* A x509 certificate for the server signed by the root CA
+* Two x509 certificates one for each of two clients, that could be used to connect to the applicaiton
+
+The root CA will use a private key that's 4096 bits and the client and server certs will use private keys with 2048 bits
+
+### Authorization
+
+To keep things simple, the following statement should hold for every process started on the server:
+
+`Only clients authenticated with the client certs having the same CN can get any information about, or interact with a certain process`
```

Comment:
that depends on the definition of 'user'. Any client using a cert with the same CN will be able to manage the process. Technically this could mean even multiple users, or a group of actors, but from the systems POV they are the same 'user'